### ISSUE TITLE
The typing of `KPoint{D}` needs to be changed

### ISSUE THREAD

**Issue by brainandforce** [brainandforce/Electrum.jl#204]
Currently, `KPoint{D}` is a subtype of `DenseVector{Float64}`. In general, [we should not be subtyping `DenseArray` at all](https://discourse.julialang.org/t/more-specific-definition-for-densearray/33591/4), and because the implementation of `KPoint{D}` is a wrapper for a `StaticArray{D,Float64}` plus a `Float64` shift parameter, we should define the following:

```julia
struct KPoint{D,T} <: StaticVector{D,T}
    point::SVector{D,T}
    weight::T
end
```
This would be a breaking change, as it would propagate to `KPointMesh{D}`, `ReciprocalDataGrid{D}`, and `PlanewaveWavefunction{D}`.

**Comment by brainandforce** [comment]
I have a better idea: a `ShiftVector{D,R<:BySpace{D},T} <: StaticVector{D,T}` type which can accomodate data shifts in either real or reciprocal space, and `KPoint{D,T}` aliases ShiftVector{D,ByReciprocalSpace{D},T}.

This should also dramatically simplify the new data grid implementation coming in 0.2.0.

**Comment by brainandforce** [comment]
This is already implemented in the `next` branch and will be resolved by #208.
